### PR TITLE
webgpu: Tune MatMulSmallOutputSizeProgram

### PR DESCRIPTION
**Pull request by qjia7**
To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6962)

### CODE REVIEW

File: tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts [hunk 1]
```diff
@@ -104,6 +104,19 @@ export function batchMatMulImpl({
...
+  // workgroups by a smaller tiling size can get better performance.
+  const matMulSmallOutputSizeProgramContidion =
+      backend.adapterInfo.isIntelGen12GPU &&
+          env().get('WEBGPU_HARDWARE_EU_NUM') as number >= 96 ?
+      Math.ceil(outerShapeA / 32) * Math.ceil(outerShapeB / 32) <= 16 :
+      (outerShapeA <= 16 &&
+       (outerShapeB <= 512 || innerShapeB >= 2 * outerShapeB)) ||
+          (outerShapeB <= 16 &&
+           (outerShapeA <= 512 || innerShapeA >= 2 * outerShapeA));
   if (matmulProgramType < 0) {
```
Comment (xhcao) [inline tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts:120]: 1. Should the defination of `matMulSmallOutputSizeProgramContidion` be in `if (matmulProgramType < 0) {...}` block?
2. Is it better to unify all conditions? Also define `matMulReduceProgramCondition`, `matMulSplitKProgramCondition`,...
3. Use `bool useMatMulSmallOutputSizeProgram = ...`?
All these are only suggestions,  must NOT be addressed.

File: tfjs-backend-webgpu/src/adapter_info.ts [hunk 2]
```diff
@@ -18,21 +18,30 @@
 export class AdapterInfo {
   private vendor: string;
   private architecture: string;
+  private intelGpuGeneration: number;
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/adapter_info.ts:21]:
```suggestion
  private intelGPUGeneration: number;
```

File: tfjs-backend-webgpu/src/adapter_info.ts [hunk 3]
```diff
@@ -18,21 +18,30 @@
...
 
   constructor(adapterInfo: GPUAdapterInfo) {
     if (adapterInfo) {
       this.vendor = adapterInfo.vendor;
       this.architecture = adapterInfo.architecture;
+      this.intelGpuGeneration = this.getIntelGpuGeneration();
     }
   }
 
+  getIntelGpuGeneration() {
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/adapter_info.ts:31]:
```suggestion
  getIntelGPUGeneration() {
```

File: tfjs-backend-webgpu/src/adapter_info.ts [hunk 4]
```diff
@@ -18,21 +18,30 @@
...
+
   isIntel(): boolean {
     return this.vendor === 'intel';
   }
 
-  isIntelGen12GPU(): boolean {
-    return this.isIntel() &&
-        (this.architecture.startsWith('gen-12') ||
-         this.architecture.startsWith('xe'));
+  isIntelGen12OrHigherGPU(): boolean {
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/adapter_info.ts:44]: Maybe we don't need this specific function.

File: tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts [hunk 5]
```diff
@@ -104,30 +104,33 @@ export function batchMatMulImpl({
...
+    // Usually increasing workgroups is a good way to gain more performance for
+    // few workgroups by tiling 32x32 (default matmul algorithm). Currently,
+    // there are three ways to increase workgroups. 1) MatMulReduceProgram,
+    // which is used only when the output size is very small (128 for now). 2)
+    // MatMulSplitKProgram, increasing workgroups by spliting K. 3)
+    // MatMulSmallOutputSizeProgram, increasing workgroups by small tile size.
+    // For different devices, the minimum optimal workgroups may be different.
+    // So here we set a |thresholdForFewWorkGroups| to indicate whether we need
+    // to increase workgroups. And the literal number is an empirical value.
+    const thresholdForFewWorkGroups =
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts:117]:
```suggestion
    const thresholdForFewWorkgroups =
```

Comment (gyagp) [reply tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts:117]: Maybe calling it threshold is enough.

File: tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts [hunk 6]
```diff
@@ -104,30 +104,33 @@ export function batchMatMulImpl({
...
+        backend.adapterInfo.isIntelGen12OrHigherGPU() &&
+            env().get('WEBGPU_HARDWARE_EU_NUM') as number >= 96 ?
+        16 :
+        8;
+    const isSmallOutputSize =
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts:122]:
```suggestion
    const hasSmallOutputSize =
```

File: tfjs-backend-webgpu/src/flags_webgpu.ts [hunk 7]
```diff
@@ -70,3 +70,8 @@ ENV.registerFlag('WEBGPU_IMPORT_EXTERNAL_TEXTURE', () => true);
  * Whether to use conv2dNaive for debugging.
  */
 ENV.registerFlag('WEBGPU_USE_NAIVE_CONV2D_DEBUG', () => false);
+
+/**
+ * Default hardware Execution units number.
+ */
+ENV.registerFlag('WEBGPU_HARDWARE_EU_NUM', () => 96);
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/flags_webgpu.ts:77]:
```suggestion
ENV.registerFlag('WEBGPU_INTEL_EU_COUNT', () => 96);
```

File: tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts (hunk 1 above)
Comment (qjia7) [reply tfjs-backend-webgpu/src/kernels/BatchMatMul_impl.ts:107]: Done.

**Review by qjia7 — COMMENTED** [review]
@gyagp  @xhcao  Please take another look, thanks.

**Review by xhcao — APPROVED** [review]
LGTM, thanks

**Review by gyagp — COMMENTED** [review]
LGTM

**Review by gyagp — APPROVED** [review]

**Pull request merged**